### PR TITLE
HDDS-4121. Implement OmMetadataMangerImpl#getExpiredOpenKeys.

### PR DESCRIPTION
**Pull request by errose28**
## What changes were proposed in this pull request?

Create an implementation to the existing OmMetadataMangerImpl#getExpiredOpenKeys method stub as the first step towards completing parent Jira HDDS-4120. This method returns a list of key information for all keys in the OM open key table whose creation time is older than a time interval configurable with the existing ozone.open.key.expire.threshold setting. The existing default value for this setting (one day) was not changed.

## What is the link to the Apache JIRA

HDDS-4121

## How was this patch tested?

The unit test TestOmMetadataManager#testGetExpiredOpenKeys was added, which creates a mixture of expired and unexpired open keys, and tests whether the expired open keys are read correctly. Functionality was added to TestOMRequestUtils to allow setting customized creation times for keys to simulate expiration.

### CODE REVIEW

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/KeyManager.java [hunk 1]
```diff
@@ -186,10 +186,11 @@ OmKeyLocationInfo allocateBlock(OmKeyArgs args, long clientID,
    * the key name and all its associated block IDs. A pending open key has
    * prefix #open# in OM DB.
    *
+   * @param count The maximum number of expired opne keys to return.
    * @return a list of {@link BlockGroup} representing keys and blocks.
    * @throws IOException
    */
-  List<BlockGroup> getExpiredOpenKeys() throws IOException;
+  List<BlockGroup> getExpiredOpenKeys(int count) throws IOException;
```
Comment (bharatviswa504) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/KeyManager.java:193]: Here, we can collect just expired KeyNames.

Comment (bharatviswa504) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/KeyManager.java:193]: And also I think we might need last expired key also, as in next iteration we need to start from that one.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/KeyManager.java [hunk 2]
```diff
@@ -186,10 +186,11 @@ OmKeyLocationInfo allocateBlock(OmKeyArgs args, long clientID,
    * the key name and all its associated block IDs. A pending open key has
    * prefix #open# in OM DB.
    *
+   * @param count The maximum number of expired opne keys to return.
```
Comment (bharatviswa504) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/KeyManager.java:189]: opne -> open

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/KeyManager.java (hunk 1 above)
Comment (bharatviswa504) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/KeyManager.java:193]: You can ignore my 2nd comment, as delete expiredKeys will be done in a single iteration, those expired keys will not be there in the openKeyTable.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/KeyManager.java [hunk 3]
```diff
@@ -186,10 +186,11 @@ OmKeyLocationInfo allocateBlock(OmKeyArgs args, long clientID,
    * the key name and all its associated block IDs. A pending open key has
    * prefix #open# in OM DB.
```
Comment (errose28) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/KeyManager.java:187]: Is this line in the original documentation still relevant? It looks like open keys are placed in the open key table unprefixed.

File: hadoop-ozone/ozone-manager/src/test/java/org/apache/hadoop/ozone/om/TestOmMetadataManager.java [hunk 4]
```diff
@@ -521,6 +527,80 @@ public void testListKeysWithFewDeleteEntriesInCache() throws Exception {
...
+    // Add unexpired keys to open key table.
+    for (int i = 0; i < numUnexpiredOpenKeys; i++) {
+      OmKeyInfo keyInfo = TestOMRequestUtils.createOmKeyInfo(volumeName,
+              bucketName, "unexpired" + i, HddsProtos.ReplicationType.RATIS,
+              HddsProtos.ReplicationFactor.ONE);
+
+      TestOMRequestUtils.addKeyToTable(true, false,
+              keyInfo, clientID, 0L, omMetadataManager);
+
+      String groupID = omMetadataManager.getOpenKey(volumeName, bucketName,
```
[the remark this answers is not in the document]
Comment (errose28) [reply hadoop-ozone/ozone-manager/src/test/java/org/apache/hadoop/ozone/om/TestOmMetadataManager.java:572]: Yeah. Removed it and pushing again.

**Comment by bharatviswa504** [comment]
Thank You @errose28 for the contribution.

**Pull request merged**